### PR TITLE
executor: add open inteferce for Executor.

### PR DESCRIPTION
**Pull request by hanfei1991**
Everytime we execute a statement, we should call open , then next, then close.

Open do some work of initialization.

@shenli @coocood @zimulala PTAL

### CODE REVIEW

File: executor/merge_join.go [hunk 1]
```diff
@@ -284,6 +282,19 @@ func (e *MergeJoinExec) Close() error {
...
+func (e *MergeJoinExec) Open() error {
+	e.prepared = false
+	e.cursor = 0
+	e.outputBuf = nil
+
+	err := e.leftRowBlock.reader.Open()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return e.rightRowBlock.reader.Open()
```
Comment (XuHuaiyu) [inline executor/merge_join.go]: errors.Trace()

File: executor/prepared.go [hunk 2]
```diff
@@ -96,11 +96,16 @@ func (e *PrepareExec) Next() (*Row, error) {
 	return nil, e.Err
 }
 
-// Close implements plan.Plan Close interface.
+// Close implements Executor Close interface.
```
Comment (XuHuaiyu) [inline executor/prepared.go]: s/Executor/the Executor

File: executor/prepared.go [hunk 3]
```diff
@@ -96,11 +96,16 @@ func (e *PrepareExec) Next() (*Row, error) {
...
 func (e *PrepareExec) Close() error {
 	return nil
 }
 
+// Open implements Executor Open interface.
```
Comment (XuHuaiyu) [inline executor/prepared.go]: ditto

File: executor/sort.go [hunk 4]
```diff
@@ -38,14 +43,15 @@ type SortExec struct {
 
 // Close implements the Executor Close interface.
 func (e *SortExec) Close() error {
-	e.fetched = false
 	e.Rows = nil
-	return e.Src.Close()
+	return e.children[0].Close()
```
Comment (XuHuaiyu) [inline executor/sort.go]: errors.Trace(err)

File: executor/sort.go [hunk 5]
```diff
@@ -38,14 +43,15 @@ type SortExec struct {
...
 }
 
-// Schema implements the Executor Schema interface.
-func (e *SortExec) Schema() *expression.Schema {
-	return e.schema
+// Open implements the Executor Open interface.
+func (e *SortExec) Open() error {
+	e.fetched = false
+	e.Rows = nil
+	return e.children[0].Open()
```
Comment (XuHuaiyu) [inline executor/sort.go]: ditto

File: executor/executor.go [hunk 6]
```diff
@@ -851,19 +772,27 @@ func (e *UnionExec) fetchData(idx int) {
...
+func (e *UnionExec) Open() error {
+	e.finished.Store(false)
+	e.resultCh = make(chan *execResult, len(e.children))
+	e.closedCh = make(chan struct{})
+	e.cursor = 0
+	var err error
+	for i, child := range e.children {
+		err = child.Open()
+		if err != nil {
+			break
```
Comment (hanfei1991) [inline executor/executor.go]: If we should run 'waitAllFinished' to avoid leak

File: executor/aggregate.go [hunk 7]
```diff
@@ -41,26 +41,26 @@ type HashAggExec struct {
...
 func (e *HashAggExec) Close() error {
-	e.executed = false
 	e.groupMap = nil
 	e.groupIterator = nil
 	for _, agg := range e.AggFuncs {
-		agg.Clear()
+		agg.Reset()
 	}
-	return e.Src.Close()
+	return e.children[0].Close()
```
Comment (lamxTyler) [inline executor/aggregate.go]: errors.Trace?

File: executor/aggregate.go [hunk 8]
```diff
@@ -41,26 +41,26 @@ type HashAggExec struct {
...
 
-// Schema implements the Executor Schema interface.
-func (e *HashAggExec) Schema() *expression.Schema {
-	return e.schema
+// Open implements the Executor Open interface.
+func (e *HashAggExec) Open() error {
+	e.executed = false
+	e.groupMap = mvmap.NewMVMap()
+	e.groupIterator = e.groupMap.NewIterator()
+	return e.children[0].Open()
```
Comment (lamxTyler) [inline executor/aggregate.go]: ditto

File: executor/aggregate.go [hunk 9]
```diff
@@ -41,26 +41,26 @@ type HashAggExec struct {
 
 // Close implements the Executor Close interface.
 func (e *HashAggExec) Close() error {
-	e.executed = false
```
Comment (lamxTyler) [inline executor/aggregate.go]: Why deleted?

File: executor/aggregate.go [hunk 10]
```diff
@@ -164,19 +156,14 @@ type StreamAggExec struct {
...
-		agg.Clear()
+		agg.Reset()
 	}
-	return e.Src.Close()
-}
-
-// Schema implements the Executor Schema interface.
-func (e *StreamAggExec) Schema() *expression.Schema {
-	return e.schema
+	return e.children[0].Open()
```
Comment (lamxTyler) [inline executor/aggregate.go]: ditto.

File: executor/distsql.go [hunk 11]
```diff
@@ -379,6 +379,13 @@ type XSelectIndexExec struct {
 	partialCount    int
 }
 
+// Open implements Exec Open interface.
```
Comment (lamxTyler) [inline executor/distsql.go]: the Executor

File: executor/join.go [hunk 12]
```diff
@@ -853,9 +865,14 @@ func (e *ApplyJoinExec) Schema() *expression.Schema {
...
 func (e *ApplyJoinExec) Close() error {
+	return nil
+}
+
+// Open implements the Executor interface.
+func (e *ApplyJoinExec) Open() error {
 	e.cursor = 0
 	e.resultRows = nil
-	return e.join.Close()
+	return e.join.Open()
```
Comment (lamxTyler) [inline executor/join.go]: ditto

File: executor/aggregate_test.go [hunk 13]
```diff
@@ -306,97 +306,6 @@ func (s *testSuite) TestAggregation(c *C) {
 	tk.MustQuery("select sum(a), sum(b) from idx_agg where b > 0 and b < 10")
 }
 
-func (s *testSuite) TestStreamAgg(c *C) {
```
Comment (lamxTyler) [inline executor/aggregate_test.go]: This test is no longer needed?

File: executor/aggregate.go (hunk 9 above)
Comment (hanfei1991) [inline executor/aggregate.go]: `Close` is for releasing resource. `Open` is for initializing.

File: executor/aggregate_test.go (hunk 13 above)
Comment (hanfei1991) [inline executor/aggregate_test.go]: This test can be covered by others. And it's inconvenient to export everything in baseExecutor.

**Comment by hanfei1991** [comment]
@XuHuaiyu @lamxTyler PTAL

File: executor/join.go [hunk 14]
```diff
@@ -518,13 +527,16 @@ func (e *NestedLoopJoinExec) Schema() *expression.Schema {
 func (e *NestedLoopJoinExec) Close() error {
 	e.resultRows = nil
 	e.innerRows = nil
+	return e.BigExec.Close()
```
Comment (XuHuaiyu) [inline executor/join.go]: close SmallExec

File: executor/join.go [hunk 15]
```diff
@@ -670,17 +682,20 @@ type HashSemiJoinExec struct {
...
-	e.prepared = false
-	e.hashTable = make(map[string][]*Row)
-	e.smallTableHasNull = false
+	e.hashTable = nil
 	e.resultRows = nil
-	err := e.smallExec.Close()
-	if err != nil {
-		return errors.Trace(err)
-	}
 	return e.bigExec.Close()
```
Comment (XuHuaiyu) [inline executor/join.go]: close smallExec

File: executor/executor.go [hunk 16]
```diff
@@ -886,35 +815,15 @@ func (e *UnionExec) Next() (*Row, error) {
...
+	<-e.closedCh
 	e.rows = nil
-	for _, sel := range e.Srcs {
-		er := sel.Close()
-		if er != nil {
-			return errors.Trace(er)
-		}
-	}
-	return nil
+	return e.baseExecutor.Close()
```
Comment (XuHuaiyu) [inline executor/executor.go]: errors.Trace()

File: executor/executor.go [hunk 17]
```diff
@@ -738,34 +666,29 @@ func (e *ExistsExec) Next() (*Row, error) {
...
+	evaluated bool
 }
 
-// Close implements the Executor Close interface.
-func (e *MaxOneRowExec) Close() error {
+// Open implements the Executor Open interface.
+func (e *MaxOneRowExec) Open() error {
 	e.evaluated = false
-	return e.Src.Close()
+	return e.children[0].Open()
```
Comment (XuHuaiyu) [inline executor/executor.go]: errors.Trace()

File: executor/executor.go [hunk 18]
```diff
@@ -696,37 +629,32 @@ func (e *TableScanExec) getRow(handle int64) (*Row, error) {
...
+	evaluated bool
 }
 
-// Close implements the Executor Close interface.
-func (e *ExistsExec) Close() error {
+// Open implements the Executor Open interface.
+func (e *ExistsExec) Open() error {
 	e.evaluated = false
-	return e.Src.Close()
+	return e.children[0].Open()
```
Comment (XuHuaiyu) [inline executor/executor.go]: errors.Trace()

File: executor/executor.go [hunk 19]
```diff
@@ -925,22 +834,17 @@ func (e *DummyScanExec) Next() (*Row, error) {
...
-	return e.schema
-}
-
-// Close implements the Executor Close interface.
-func (e *CacheExec) Close() error {
+// Open implements the Executor Open interface.
+func (e *CacheExec) Open() error {
 	e.cursor = 0
-	return nil
+	return e.children[0].Open()
```
Comment (XuHuaiyu) [inline executor/executor.go]: ditto

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**